### PR TITLE
Public Authorized SSH Keys

### PR DESCRIPTION
**Pull request by dperpeet**
Replaces #2520

### CODE REVIEW

File: pkg/shell/authorized-keys.js [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+                self.state = "failed";
+                console.warn("Error proccessing authentication keys: "+ ex);
+            }
+            $(self).triggerHandler("changed");
+        }
+
+        function fix_permissions() {
+            // fix permissions, in case we created the file as root
+            cockpit.spawn(["chown", "-R", user_name+":"+user_name, dir],
+                           {'superuser': 'try'})
```
Comment (stefwalter) [inline pkg/shell/authorized-keys.js]: This assumes that the user has a primary group with the same name. I would suggest avoiding that assumption. The very next line sets ```chmod 600``` which makes the group irrelevant.

In addition this has escaping issues concatenating the various values.

File: pkg/shell/authorized-keys.js [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+                            df.resolve(obj);
+                        });
+                })
+                .fail(function(ex) {
+                    df.reject(ex);
+                });
+            return df;
+        }
+
+        function parse_keys (content, tag, ex) {
```
Comment (stefwalter) [inline pkg/shell/authorized-keys.js]: Extra space after function name.

File: pkg/shell/authorized-keys.js [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+                    tmp_file = cockpit.file(tmp_file_path);
+                    process_next();
+                })
+                .fail(process_failure);
+        }
+
+        self.add_key = function(key) {
+            var df = $.Deferred();
+            key = key.trim();
+            cockpit.spawn(["mktemp"])
```
Comment (stefwalter) [inline pkg/shell/authorized-keys.js]: Why are we doing so many round trips here. I would really suggest just passing in a shell script, with the key as input, and performing all these actions from the script.

Comment (stefwalter) [inline pkg/shell/authorized-keys.js]: Hmmm, it seems that the iterative modifications require that it's done this way. But even then the parsing of the key could be done in a single step and not three.

File: pkg/shell/authorized-keys.js [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+                    update_keys(keys, tag);
+                } else {
+                    process_next();
+                }
+            }
+
+            function process_next() {
+                var raw = lines[processed];
+                if (previous[raw] === undefined) {
+                    create_key_object (raw, tmp_file, tmp_file_path)
```
Comment (stefwalter) [inline pkg/shell/authorized-keys.js]: Ditto here, it seems we're doing a hundred round trips just to parse the file.

**Comment by stefwalter** [comment]
Seems like blank lines say *Invalid Key*

screenshot from 2015-08-03 10-56-11

**Comment by stefwalter** [comment]
We should handle blank lines in the way that ```sshd``` does. In other words, if it ignores blank lines, or lines with comments, so should we.

**Comment by stefwalter** [comment]
This branch does not match the design: https://github.com/cockpit-project/cockpit/wiki/Feature:-Key-based-authentication

**Comment by stefwalter** [comment]
@dperpeet 4 commits here that should fix parsing, add tests, and bring this in line with the wireframes. https://github.com/stefwalter/cockpit/commits/keys

**Comment by dperpeet** [comment]
The problem for tests is that older systems print the fingerprint in md5 representation, while the default in newer versions was switched to sha256 ([example](https://bbs.archlinux.org/viewtopic.php?id=196274))
At least Jenkins and F21 appear to be affected. There doesn't seem to be a way to force the output of older versions to be in the new format.

**Comment by dperpeet** [comment]
after #2533 was merged, shell/Makefile.am needs to be adapted to include the new .min.js files.

**Comment by stefwalter** [comment]
I'll normalize the fingerprints during the testing.

**Comment by stefwalter** [comment]
Two more commits on my stefwalter/keys branch should fix the problems here: https://github.com/stefwalter/cockpit/commits/keys

**Comment by stefwalter** [comment]
I pushed another FIXUP to my keys branch.

**Comment by dperpeet** [comment]
> Hmmm, not sure why we're using /bin/bash here. But the web interface uses /bin/sh with the scripts, and we should do the same when testing the scripts.

jenkins ci uses some other shell and the tests fail if we don't use bash

**Pull request closed without merging**